File: daily-challenge/2025-09-17/solution.py
"""
Problem: 2345 - Design a Food Rating System  
Description: Design a system that supports initializing with lists of foods, their cuisines, and ratings, changing a food's rating, and querying the highest-rated food for a given cuisine (ties broken by lexicographically smaller name).  
Date: 2025-09-17
"""

import heapq
from collections import defaultdict
from typing import List

class FoodRatings:
    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        self.food_to_cuisine = {}
        self.food_to_rating = {}
        self.cuisine_heaps = defaultdict(list)  # cuisine -> list of (-rating, food)

        for f, c, r in zip(foods, cuisines, ratings):
            self.food_to_cuisine[f] = c
            self.food_to_rating[f] = r
            heapq.heappush(self.cuisine_heaps[c], (-r, f))

    def changeRating(self, food: str, newRating: int) -> None:
        c = self.food_to_cuisine[food]
        self.food_to_rating[food] = newRating
        heapq.heappush(self.cuisine_heaps[c], (-newRating, food))

    def highestRated(self, cuisine: str) -> str:
        heap = self.cuisine_heaps[cuisine]
        while heap:
            neg_r, name = heap[0]
            if self.food_to_rating.get(name, None) == -neg_r:
                return name
            heapq.heappop(heap)
        return ""  # per problem constraints this line should never be reached

File: daily-challenge/2025-09-17/solution.cpp
"""
Problem: 2345 - Design a Food Rating System  
Description: Design a system that supports initializing with lists of foods, their cuisines, and ratings, changing a food's rating, and querying the highest-rated food for a given cuisine (ties broken by lexicographically smaller name).  
Date: 2025-09-17
"""

#include <bits/stdc++.h>
using namespace std;

class FoodRatings {
    unordered_map<string, string> foodToCuisine;
    unordered_map<string, int> foodToRating;
    unordered_map<string, set<pair<int,string>>> cuisineSets; // (-rating, name)

public:
    FoodRatings(vector<string>& foods, vector<string>& cuisines, vector<int>& ratings) {
        int n = foods.size();
        for (int i = 0; i < n; ++i) {
            const string &f = foods[i];
            const string &c = cuisines[i];
            int r = ratings[i];
            foodToCuisine[f] = c;
            foodToRating[f] = r;
            cuisineSets[c].insert({-r, f});
        }
    }

    void changeRating(string food, int newRating) {
        string c = foodToCuisine[food];
        int old = foodToRating[food];
        auto &s = cuisineSets[c];
        s.erase({-old, food});
        s.insert({-newRating, food});
        foodToRating[food] = newRating;
    }

    string highestRated(string cuisine) {
        auto &s = cuisineSets[cuisine];
        return s.begin()->second;
    }
};
